ocean_lib/data_provider/data_service_provider.py

```diff
@@ -205,33 +187,47 @@ def download_service(
             indexes = [index]
 
         req = PreparedRequest()
-        params = {
-            "documentId": did,
+        service_endpoint += f"?documentId={did}"
+        payload = {
             "serviceId": service_id,
-            "serviceType": ServiceTypes.ASSET_ACCESS,
-            "dataToken": token_address,
-            "transferTxId": order_tx_id,
-            "consumerAddress": wallet.address,
+            "consumerAddress": consumer_wallet.address,
+            "transferTxId": tx_id,
         }
 
         if userdata:
             userdata = json.dumps(userdata)
-            params["userdata"] = userdata
-
-        req.prepare_url(service_endpoint, params)
-        base_url = req.url
+            payload["userdata"] = userdata
 
         for i in indexes:
-            signature = DataServiceProvider.sign_message(wallet, did)
-            download_url = base_url + f"&signature={signature}&fileIndex={i}"
-            logger.info(f"invoke consume endpoint with this url: {download_url}")
-            response = DataServiceProvider._http_method(
-                "get", download_url, stream=True
+            payload["fileIndex"] = i
+            payload["nonce"], payload["signature"] = DataServiceProvider.sign_message(
+                consumer_wallet, did
             )
+            req.prepare_url(service_endpoint, payload)
+            response = DataServiceProvider._http_method("get", req.url)
+
+            if not response or not hasattr(response, "status_code"):
+                raise Exception("Response not found!")
+
+            if response.status_code != 200:
+                msg = (
+                    f"Download asset failed at the downloadEndpoint "
+                    f"{req.url}, reason {response.text}, status {response.status_code}"
+                )
+                logger.error(msg)
+                raise Exception(msg)
```

Comment:
we do not raise a plain Exception. Never, ever. Please see my other comment.

Reply:
I recycled some chunks of the previous code from V3.